src/java.net.http/share/classes/jdk/internal/net/http/Http1Exchange.java

```diff
@@ -68,11 +69,11 @@
 
     /** Records a possible cancellation raised before any operation
      * has been initiated, or an error received while sending the request. */
-    private Throwable failed;
+    private AtomicReference<Throwable> failedRef = new AtomicReference<>();
```

Comment:
Seems it's only modified under lock. Can we use `volatile` instead?

Reply:
I prefer to keep AtomicReference - it's easier to reason about.
Maybe at some time I'll do a pass to use VarHandles in various places but now is not the time.

Reply:
let's make it `final` at least

Reply:
Oops! Thanks for noticing that. That's a bug. Done.